### ISSUE TITLE
Make warn_on_inefficient_map check dtype isn't Object before warning

### ISSUE THREAD

**Issue by deanm0000** [pola-rs/polars#15587]
### Description

This is inspired by [this](https://github.com/pola-rs/polars/issues/15582)

Ideally we wouldn't warn about inefficient map when operating on an Object column since that's the only way to do anything with it. I poked around the source a bit to see if I could do it semi-quickly but I see that where it is that it's non-trivial to get the column's dtype.

One thing I can't figure out is that if I do

```
df=pl.DataFrame({'a':[uuid4() for _ in range(10)]})
df.with_columns(pl.col('a').map_elements(lambda x: str(x)))
```

then I get the warning but if I do

```
a=pl.Series([uuid4() for _ in range(10)])
a.map_elements(lambda x: str(x))
```

then I don't get the warning. The code seems to be in both the Expr and Series version of map_elements.

Point being, if we defer the check until it's dispatched as a Series then the `.dtype` is checkable and we can skip warning when the dtype is an Object.

**Comment by alexander-beedie** [comment]
> One thing I can't figure out ... // ... then I don't get the warning

Apparently it's an accident, heh. We check that a valid column/series names is present before disassembling the lambda - in the example above the Series name is the empty string and that currently sneaks through without being checked (which I'll fix shortly as that isn't intended) :)

Unfortunately we can't know the dtype of the input at the point the lambda is analysed; I think our best bet is to bite the bullet and get around to enabling Object → String casts 🤔

**Comment by deanm0000** [comment]
In the Series version we can b/c a Series carries around it's dtype. In that way the check can be deferred until it hits the Series method but then there needs to be another means of keeping track of the need to check from map_elements to map_batches. To your point, it's not low hanging fruit. The low hanging fruit is the Object to String explicit cast. I'll leave this open b/c it's relevant for other things than just casting to string. That was just the example.

**Comment by alexander-beedie** [comment]
> In the Series version we can b/c a Series carries around it's dtype. In that way the check can be deferred until it hits the Series method but then there needs to be another means of keeping track of the need to check from map_elements to map_batches. To your point, it's not low hanging fruit. The low hanging fruit is the Object to String explicit cast. I'll leave this open b/c it's relevant for other things than just casting to string. That was just the example.

Yup, I should have clarified; the _expression_ API doesn't know - you're quite right that we can pass in the dtype from Series usage, and I'll probably have a look at doing so. However, we don't want to defer analysis of the function until later. In the ideal case the user is immediately alerted _while_ they're constructing the query interactively, and can act to modify it then & there, rather than having a warning appear sometime later in  pipeline (which may not be as easy to identify and act on). 

So, I'm ok with not catching 100% of cases as long as we get a good proportion of them at the most opportune time. I'm not sure how much it would get used, but we could perhaps add an optional "input_dtype" to the expression variant of `map_elements` for those who want to got the extra mile and provide the information at point of use? Obviously not as nice as taking a cue from Nike and "just doing it", but possibly useful 🤔

**Comment by deanm0000** [comment]
Sorry I had a brain fart. I meant defer the check until it gets to rust (not Series) when it knows the datatype. I'm not sure how much trouble that is but I'd certainly guess more than it's worth.

**Comment by stinodego** [comment]
This would also fix https://github.com/pola-rs/polars/issues/10269

**Comment by CaselIT** [comment]
tripped on this as well while trying to migrate some code to polars. In my case the suggestion it gives
> Replace this expression...
>  - srs.map_elements(lambda v: ...)
> with this one instead:
>  + s + 1

is wrong since if I try it it leads to `TypeError: cannot do arithmetic with Series of dtype: Object and argument of type: 'int'`
